lib/Optimizer/Transforms/ObserveAnsatz.cpp

```diff
@@ -90,6 +98,29 @@ struct AnsatzFunctionAnalysis {
         data.qubitValues.insert({op.getConstantIndex(), op.getResult()});
     });
 
+    // If mapping has moved qubits or introduced auxillary qubits, update the
+    // analysis accordingly.
+    if (auto mappingAttr =
+            dyn_cast_if_present<ArrayAttr>(funcOp->getAttr("mapping_v2p"))) {
+      // First populate mapping_v2p[].
+      SmallVector<int64_t> mapping_v2p(mappingAttr.size());
+      for (auto [origIx, mappedIx] : llvm::enumerate(mappingAttr)) {
+        if (auto val = dyn_cast<IntegerAttr>(mappedIx))
+          mapping_v2p[origIx] = val.getInt();
+        else
+          emitError(funcOp.getLoc(), "invalid mapping_v2p found in function");
```

Comment:
Should this be a show stopping fatal error?

Reply:
Yes, probably. I can update the surrounding logic to fail the pass.

Reply:
Updated in ec9ad891